processing/src/main/java/io/druid/query/PerSegmentOptimizingQueryRunner.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query;
+
+import io.druid.java.util.common.guava.Sequence;
+
+import java.util.Map;
+
+/**
+ * This runner optimizes queries made on a single segment, using per-segment information,
+ * before submitting the queries to the base runner.
+ *
+ * Example optimizations include adjusting query filters based on per-segment information, such as intervals.
+ *
+ * This query runner should only wrap base query runners that will
+ * be used to query a single segment (i.e., when the query reaches a historical node).
+ *
+ * @param <T>
+ */
+public class PerSegmentOptimizingQueryRunner<T> implements QueryRunner<T>
```

Comment:
I was having a moment of worry about the naming of these types maybe being too close to`BySegmentQueryRunner` perhaps causing unnecessary confusion, but I can't really think of anything better. Naming of `QueryRunner` types is already sort of all over the place, so at least this one has javadocs :+1: